keps/sig-node/20181106-in-place-update-of-pod-resources.md

```diff
@@ -332,6 +417,19 @@ Status.ContainerStatuses[i].Resources to match the desired limit values.
   determine if in-place resize is possible.
 * Impact of memory-backed emptyDir volumes: If memory-backed emptyDir is in
   use, Kubelet will clear out any files in emptyDir upon Container restart.
+<<[UNRESOLVED]>>
```

Comment:
We have not talked about this point yet.  Thoughts?

Reply:
I apologize for delay in getting to this, i'm running two other projects.

Now that we have a working implementation with resize policies, I was able to quickly try it with two container podspec - and a resize policy so that one of the containers restarts on memory update.

As it turns out, we don't need to do anything:
  - It appears we get a tmpfs half the DRAM on the system (regardless of sizeLimit specified - bug?).
  - Container restarts for resizing, my code handles updating pod level cgroup memory.limit_in_bytes in the correct order with container restart depending on net-mem increase or decrease.
  - Restarted container comes back up with new desired memory, and tmpfs mounts has pre-restart files available.

This is exactly how you described and as per documented behavior. Now, if runtime restarts container to resize, no problem - same behavior as above where policy asked for restart. We don't tear down the pod for container restarts.

Based on my observation, we should remove this line or clarify that memory-backed emptydirs are getting their FS separate from memory.limits, so container restarts won't have any impact to how things currently work.

CC: @derekwaynecarr Please let me know if I missed something. I recall you mentioned there are challenges where the charge gets transferred to pod level cgroup on container stop. I might be missing something, but didn't see any relevant cgroup parameter specifying charge - but frankly i'm not sure how or why the mounts tell me 16G when I specified 2G sizeLimit so I don't exactly understand this. I'll find some time to dig into it later this week, please shoot me any pointers you may have.

Reply:
I think you are right and this can be removed.

Aside: memory attribution for tmpfs is (was?) a disaster (last I looked).  

Consider a 1-container pod, the container asks for 1GiB, and a tmpfs.  It writes 900 MiB to tmpfs.  Those pages are faulted in the container's cgroup.  If that process dies and the container cgroup is deleted, those pages are still dirty and anonymous, so they get shifted up to the pod's cgroup. Now the container restarts, and allocates 125 MiB.  The container cgroup is far below its limit, but it is OOM.  Now make that 2 containers and figure out what happened.

As long as we don't destroy the container cgroup, we're OK, I think.

Reply:
> Aside: memory attribution for tmpfs is (was?) a disaster (last I looked).

I did some experimenting - I hope I didn't miss anything and you find this useful:

- About my earlier observation of 16G tmpfs for pod spec with sizeLimit 2G, it is just what df utility shows. The pod's tmpfs mount has been given 16G (half of 32G system RAM) due to kernel default.
- Writing files that exceed sizeLimit causes pod eviction - so sizeLimit is enforced. Here, I had container memory request & limit well above sizeLimit.
- With sizeLimit above sum(container requests/limits):
  - One container pod:
    - Writing a file to use up its memory.limit causes container OOM kill as expected, transferring charge to pod.
    - Restarted container only has whatever little is left for the pod even though its cgroup has been cleared - can't even start a bash shell session.
      - I believe pod level cgroup is gating this. So a pod cannot 'exit 1' and game the system to get a clean slate of cgroup quota.
  - Two containers pod:
    - Writing a file to use up its memory.limit causes container OOM kill as above, transferring charge to pod.
    - Restarted container gets a clean cgroup slate, and can keep using emptydir/memory beyond what is rightfully its. But it is stealing from its sibling containers until it breaches aggregate pod cgroup limit. At this point both containers get the boot.

To me, this looks reasonable. Just as OS doesn't strict-police threads in a process, if two or more containers in a pod can't play nice - it is the pod owner's problem. We are holding the **pod** to what it was promised.

Now, if we introduce vertical scaling, we can force size-down a container below usage, and restart will give it a clean slate. However, the pod cgroup downward resize fails due to usage > limit. (Note: VPA should never recommend/apply a memory lower than usage)
  - We need to either proactively error out such a resize, or set limits just above usage as Derek had suggested - though it may not help the emptyDir.memory scenario.

> As long as we don't destroy the container cgroup, we're OK, I think.

Well we do have a problem .. we don't have a CRI way to tell container to restart itself retaining its cgroup or whatever resource control mechanism it uses. We stop+start, and stop kills the cgroup at-least with containerd runtime, and how runtime implements stop is beyond our scope of influence - it is not even our concern.

This opens up a path for gaming the system if we don't validate-block such a downward resize. I suppose we could alternately force pod restart if pod cgroup resize fails, and possibly block emptyDir(memory) + policy.restart in validation .. a user could unwittingly walk into this unpleasant surprise. I need some time to think about this more..

Reply:
I think it's not "gaming" the system.  As you pointed out - the pod is still constrained.  It would be hard to diagnose, for sure, and it leaves room for improvement, but it isn't WRONG.

Short of becoming much more prescriptive about cgroup management, it might be about as good as we can get.

Reply:
Fair enough. We can resolve this for now, and always revisit should a problem arise in practice.